### DISCUSSION TITLE
[Eldoc] Disable eldoc in message area, and only show doc details when I invoke eldoc-doc-buffer?

### DISCUSSION THREAD

**Discussion by ianyepan** [joaotavora/eglot#691] [Q&A] [answered]
The current Eglot default enables eldoc and constantly shows the signature/details of the variable/function under my cursor, which are often multiple lines. This results in a visual distraction IMHO. Especially when I'm moving my cursor around and the echo area just flashes in and out with different documentation. Screenshot:

![image](https://user-images.githubusercontent.com/36686900/117682671-e30f7400-b1e5-11eb-8b4b-8a3966ba1988.png)

However, I love the functionality when I can invoke eldoc-doc-buffer manually (currently bound to K) and show the details of function/variable under point in a side buffer, such as:

![image](https://user-images.githubusercontent.com/36686900/117682971-2ff34a80-b1e6-11eb-9441-10ff262cb27f.png)

I was wondering if I can suppress the automatic-showing-in-echo-area-behavior, and only show details when I manually invoke eldoc-doc-buffer? Currently, the only solution I can think is to set eldoc-idle-delay to a massive number, so it's never reached. E.g.
```emacs-lisp
(setq eldoc-idle-delay most-positive-fixnum)
```
Thought this solution feels like an ugly hack...

**Comment by joaotavora** [comment, marked as answer]
I think you can try, if using the latest version of `eldoc.el`.

 `(setq eldoc-echo-area-prefer-doc-buffer t)`

**Reply by ianyepan** [reply]
@joaotavora Thanks for the speedy reply man! Sadly, this suggestion doesn't exactly solve the problem: "`eldoc-echo-area-prefer-doc-buffer: ElDoc will skip showing documentation in the echo area if the dedicated documentation buffer (given by ‘eldoc-doc-buffer’) is being displayed in some window.`"
Which means that the echo area won't show only if a doc buffer window is open. My desired behavior is for eldoc to "never" show automatically. I'd only like it to show when I manually invoke doc-buffer.

**Reply by joaotavora** [reply]
In that case, I _think_ you can turn off `eldoc-mode` and use `M-x eldoc` when you want to show documentation for the thing at point.

**Reply by ianyepan** [reply]
Thanks @joaotavora , invoking eldoc (or eldoc-doc-buffer) works as expected even when eldoc-mode is turned off :-) And with eldoc-mode off, those echo buffers don't show up anymore.

My last question is: Since loading eglot automatically turns on eldoc-mode in the local buffer, me simply adding `(global-eldoc-mode -1)` in my init.el would not be enough. I was thinking about adding a hook to disable eldoc-mode AFTER eglot is loaded. What would be a good way to achieve this?

**Reply by joaotavora** [reply]
I think there is a pull request/issue somewhere here to make `eldoc` be exempted from Eglot's auto-setting using `eglot-stay-out-of`.  I seem to have lost it, so if you can search for it and bring it to my attention, I'd be thankful.

Meanwhile, maybe you can put `(eldoc-mode -1)` in `eglot-managed-mode-hook`

**Reply by ianyepan** [reply]
Cheers @joaotavora , **"eglot-managed-mode-hook" was the variable I was looking for!** The following configuration solved my problem:
```emacs-lisp
(setq eglot-managed-mode-hook '((lambda () (eldoc-mode -1))))
```
And manually invoking eldoc-doc-buffer still works as expected with or without eldoc-mode being enabled! 😊

**Reply by joaotavora** [reply]
No problem, though you should be using a variation on this which is more syntactically correct:

```
(setq eglot-managed-mode-hook (list (lambda () (eldoc-mode -1))))
```

Be sure to mark this solved.

**Reply by ianyepan** [reply]
> you should be using a variation on this which is more syntactically correct

Thanks @joaotavora ! 

Marked as solved.

**Reply by ianyepan** [reply]
@joaotavora Not to say I have a better understanding in Elisp (I most certainly don't), but doesn't these two lines mean the same thing?
```emacs-lisp
(setq eglot-managed-mode-hook (list (lambda () (eldoc-mode -1))))
; v.s.
(setq eglot-managed-mode-hook '((lambda () (eldoc-mode -1))))
```

**Reply by joaotavora** [reply]
They look similar and both work, but the second implies a kind of `eval` at runtime, and prevents the compiler from seeing the insides of the lambda as code (it sees it simply as data) destroying the possible optimizations, etc.    Doesn't make much difference here, but don't put lambdas inside quoted lists, unless you know very well what you are doing.

Here, you should use `add-hook` (forgot to suggest that):

```
(add-hook 'eglot-managed-mode-hook (lambda () (eldoc-mode -1)))
```

**Reply by ianyepan** [reply]
> Here, you should use add-hook (forgot to suggest that)

Cheers, I'll go with this one then.

**Comment by dgutov** [comment]
First of all, there is a bug in Eglot: it doesn't honor the default value of `eldoc-echo-area-use-multiline-p` appropriately.

@ianpan870102 Try `(setq eldoc-echo-area-use-multiline-p nil)` first. That should stop the echo area from resizing and windows from jumping. But still leave a little blinking down there.

Regarding the latter, how would you like if the signature hints *did* show up in the echo area, but the "hover" tips (like the doc of the identifier at point) did not? All of the remaining blinking seems to come from the latter, and I'm not sure how easy it is to fix in Eglot.

**Reply by ianyepan** [reply]
> Try (setq eldoc-echo-area-use-multiline-p nil) first. That should stop the echo area from resizing and windows from jumping. But still leave a little blinking down there.

Thanks for the tip. But personally I prefer to have the echo area completely clean (just personal preference).

> how would you like if the signature hints did show up in the echo area, but the "hover" tips (like the doc of the identifier at point) did not? All of the remaining blinking seems to come from the latter.

Same response as above -- I don't think the echo area is a great place to show signature/hints/doc overall. But I see your point -- that'd be a great option for users who find the echo area valuable, yet don't want to be distracted by multi-line docs.

**Reply by joaotavora** [reply]
> First of all, there is a bug in Eglot: it doesn't honor the default value of eldoc-echo-area-use-multiline-p appropriately.

Make a bug report, follow the guidelines, and thanks in advance.

**Reply by dgutov** [reply]
> But personally I prefer to have the echo area completely clean (just personal preference).

:ok_hand: 

> I don't think the echo area is a great place to show signature/hints/doc overall

It would be better to show the signature hint somewhere closer to the cursor indeed. But we don't have a standard UI for that yet.

**Reply by dgutov** [reply]
@joaotavora The information is all here already.

**Reply by joaotavora** [reply]
Perhaps to you and your LSP setup: if you see a bug you should be able to reproduce it, and all I'm asking you is that you do that so I reproduce it myself can prioritize its solution. It's easy, just state what servers you are using, provide a simple file, etc.

**Reply by joaotavora** [reply]
The instructions are all there in the template.

**Reply by ianyepan** [reply]
@dgutov 
>All of the remaining blinking seems to come from the latter, I'm not sure how easy it is to fix in Eglot.

I actually have an idea. Check out the entire eldoc buffer for the classic C language function `calloc()` (I'm using clangd):

```text
calloc: function calloc

---
→ void *  
Parameters:  
- size_t __nmemb
- size_t __size

Allocate NMEMB elements of SIZE bytes each, all initialized to 0.  

---
extern void *calloc(size_t __nmemb, size_t __size)
```
Or another classic C language function `fgetc()`'s complete eldoc doc:

```text
fgetc: function fgetc

---
→ int  
Parameters:  
- FILE * __stream

Read a character from STREAM.  
These functions are possible cancellation points and therefore not marked with \_\_THROW.  

---
extern int fgetc(FILE *__stream)
```

At least for the clangd LSP server, the two example above always put the single line function signature at the LAST LINE. If we want to, we could probably extract the last line and use that in the echo area only.

The problem is, every LSP server return their full doc in a different style (potentially different order. The solution that works for clangd (extracting the last line) may not work for another server, say Python's pyright, or JS's ts-ls.

**Reply by dgutov** [reply]
@joaotavora If it's really that non-obvious, I'll try to make a video of it sometime later.

**Reply by joaotavora** [reply]
You can make the video and attach it, but be sure to also provide the recipe according to the instructions. Thanks.

**Reply by joaotavora** [reply]
And yes, it's pretty non-obvious to me.  You don't even say what values aren't honoured and exactly how. This matter was studied with input from a fair number of developers and tested.  But let's save any discussion for when we have your bug report and your video.

**Reply by dgutov** [reply]
@ianpan870102 That's what I was talking about: we can show the signature only. Instead of the other code snippet, try: 

```el
(add-hook 'eglot-managed-mode-hook
          (lambda ()
            (setq eldoc-documentation-functions '(eglot-signature-eldoc-function))))
```

**Reply by ianyepan** [reply]
Just tried this approach @dgutov and it's "almost" perfect -> except that there is one downside to it: By setting `(setq eldoc-documentation-functions '(eglot-signature-eldoc-function))`, the cursor-hovering only triggers the echo area message when the cursor is placed "within" the argument list itself, and will not trigger a signature if the cursor is place on the function name.

For instance (Javascript example):
```javascript
fetch(baseURL + '/v2/countries/' + countryName)
```
Placing the cursor on `fetch` no longer has an effect. But placing the cursor inside the parentheses (e.g. on baseURL, or on v2 etc.) will generate the following docstring:
```text
fetch(input: RequestInfo, init?: RequestInit): Promise<Response>
```

**Reply by ianyepan** [reply]
It'd be awesome if placing the cursor on `fetch`, the method name, could also generate
```text
fetch(input: RequestInfo, init?: RequestInit): Promise<Response>
```

**Reply by joaotavora** [reply]
@ianpan870102 only try this approach if you're OK with completing losing the information given by `eglot-hover-eldoc-function`, which is the usual "hover on anything" information given by LSP servers.  If you are, it's a fine approach.

If you want to keep both elements in `eldoc-documentation-functions` but only display some information (presumably the one-liners given by `eglot-signature-eldoc-function`) in the echo area, you need to tweak `eldoc-display-functions` instead.

**Reply by joaotavora** [reply]
> It'd be awesome if placing the cursor on fetch, the method name, could also generate

That's completely up to the LSP server.  Eglot, the LSP client, cannot change when documentation is generated, only how it is displayed to the user.

**Reply by ianyepan** [reply]
> you need to tweak eldoc-display-functions instead.

I can't find that specific variable (`eldoc-display-functions`), I searched throughout eldoc.el.gz, and could only find `eldoc-documentation-function`)

**Reply by dgutov** [reply]
@ianpan870102 See, there are two different hints which Eglot combines into one here:

* Identifier "on hover" hint, which describes the thing under cursor (or at point, in Emacs's case). The thing can be a constant/class/module, or a method reference.
* Signature help, which helps you know what arguments to write in the current function call. That one is shown when you are in the process of writing said arguments, that is, inside the parentheses.

I suppose Eglot could be a little smarter and, when outside of any parens, would first jump inside parens and request signature help from there. First, that would be non-standard behavior (and Eglot doesn't really know what constitutes "parens" in every supported language), but second, it would be inconsistent. Consider the case of nested calls:

```js
foo(
  bar(1, 2, 3),
  qux
)
```

When your point is at `bar` (e.g. at the letter `b`), which function should you get "signature help" for? The standard behavior is to show it for `foo`, which I think makes sense.

**Reply by joaotavora** [reply]
If you are indeed using the latest Eldoc (as installed by `package-install`, I think it's 1.11.0), it won't be `eldoc.el.gz`, but rather in another overriding copy of `eldoc.el`.  There you will find this 👇 .  The tweak to perform to this is more complex, and involves replacing its first argument `eldoc-display-in-echo-area`, with something else, call it `ianpan87-display-but-only-sometimes` that filters out "unwanted" elements in DOCS  and then calls `eldoc-display-in-echo-area` itself on the cleaned up list.

```
eldoc-display-functions is a variable defined in `eldoc.el'.

Its value is (eldoc-display-in-echo-area eldoc-display-in-buffer)

  This variable may be risky if used as a file-local variable.
  Probably introduced at or before Emacs version 28.1.

Hook of functions tasked with displaying ElDoc results.
Each function is passed two arguments: DOCS and INTERACTIVE. DOCS
is a list (DOC ...) where DOC looks like (STRING :KEY VALUE :KEY2
VALUE2 ...).  STRING is a string containing the documentation's
text and the remainder of DOC is an optional list of
keyword-value pairs denoting additional properties of that
documentation.  For commonly recognized properties, see
`eldoc-documentation-functions'.

INTERACTIVE says if the request to display doc strings came
directly from the user or from ElDoc's automatic mechanisms'.

[back]
```

**Reply by ianyepan** [reply]
> latest Eldoc (as installed by package-install, I think it's 1.11.0

Oh thanks, didn't realize that. 
Potentially a dumb question: how can I override the older eldoc with this newer eldoc I just installed? Emacs seems to be picking the older eldoc still, because when I C-h v some-eldoc-var, or C-h f some-eldoc-func, I'm always guided to the eldoc.el.gz version.

**Reply by joaotavora** [reply]
If you install `Eglot` according to the instructions it should happen automatically. From here on, it comes down to your setup, some people use `use-package`, `straight` and all kinds of fancy installers.  You'll have to provide your setup, and potentially trim it down to the simplest possible where that problem still occurs. You can make a new issue, or a new conversation for just this topic, or you can ask that question in the Emacs bug tracker (I think it's drifting in this discussion topic).

**Reply by ianyepan** [reply]
Gotcha, but I think I'll stop here -- I don't have dire need to tweak eldoc-display-functions at the moment, so I'm okay with the default version of eldoc.
